haystack/nodes/other/route_documents.py

```diff
@@ -38,43 +51,89 @@ def __init__(self, split_by: str = "content_type", metadata_values: Optional[Lis
+            raise ValueError("Provide metadata_values if you want to split a list of Documents by a metadata field.")
 
     @classmethod
     def _calculate_outgoing_edges(cls, component_params: Dict[str, Any]) -> int:
         split_by = component_params.get("split_by", "content_type")
         metadata_values = component_params.get("metadata_values", None)
+        return_remaining = component_params.get("return_remaining", False)
+
         # If we split list of Documents by a metadata field, number of outgoing edges might change
         if split_by != "content_type" and metadata_values is not None:
-            return len(metadata_values)
-        return 2
+            num_edges = len(metadata_values)
+        else:
+            num_edges = 2
+
+        if return_remaining:
+            num_edges += 1
+        return num_edges
+
+    def _split_by_content_type(self, documents: List[Document]) -> Dict[str, List[Document]]:
+        mapping = {"text": "output_1", "table": "output_2"}
+        split_documents: Dict[str, List[Document]] = {"output_1": [], "output_2": [], "output_3": []}
+        for doc in documents:
+            output_route = mapping.get(doc.content_type, "output_3")
+            split_documents[output_route].append(doc)
+
+        if not self.return_remaining:
+            # Used to avoid unnecessarily calculating other_content_types depending on logging level
+            if logger.isEnabledFor(logging.WARNING) and len(split_documents["output_3"]) > 0:
+                other_content_types = {x.content_type for x in split_documents["output_3"]}
+                logger.warning(
+                    "%s document(s) were skipped because they have content type(s) %s. Only the content "
+                    "types 'text' and 'table' are routed.",
+                    len(split_documents["output_3"]),
+                    other_content_types,
+                )
+            del split_documents["output_3"]
+
+        return split_documents
+
+    def _get_metadata_values_index(self, metadata_values: Union[List[str], List[List[str]]], value: str) -> int:
+        for idx, item in enumerate(metadata_values):
+            if isinstance(item, list):
+                if value in item:
+                    return idx
+            else:
+                if value == item:
+                    return idx
+        return len(metadata_values)
+
+    def _split_by_metadata_values(
+        self, metadata_values: Union[List, List[List]], documents: List[Document]
+    ) -> Dict[str, List[Document]]:
+        split_documents: Dict[str, List] = {f"output_{i + 1}": [] for i in range(len(metadata_values) + 2)}
+        remaining_key = f"output_{len(metadata_values) + 1}"
+
+        for doc in documents:
+            current_metadata_value = doc.meta.get(self.split_by, remaining_key)
+            index = self._get_metadata_values_index(metadata_values, current_metadata_value)
+            split_documents[f"output_{index + 1}"].append(doc)
```

Comment:
I find this part a bit convoluted, can we manage to simplify it a bit? Does something like this sound good?

```suggestion
        # We need also to keep track of the excluded documents so
        # we add 2 to the number of metadata_values
        output_keys = [f"output_{i}" for i in range(1, len(metadata_values) + 2)]
        split_documents = {k: [] for k in output_keys}
        # This is the key used for excluded documents
        remaining_key = output_keys[-1]

        for doc in documents:
            current_metadata_value = doc.meta.get(self.split_by, remaining_key)
            index = self._get_metadata_values_index(metadata_values, current_metadata_value)
            output = output_keys[index]
            split_documents[output].append(doc)
```

Reply:
Yes I agree this looks good and is much more readable.